tests/tracer/test_provider.py

```diff
@@ -0,0 +1,218 @@
+
+
+def test_default_cross_execution_parenting_multi_span_close_wrong_executor(tracer, test_spans):
+    """
+    When an active trace is continued in a new executor
+        When the propagated span is closed in the wrong executor
+            Spans created in the new executor should inherit from the previous
+            executor without issue
+    """
+
+    provider = TestContextProvider()
+    tracer.configure(context_provider=provider)
+
+    main_span = tracer.trace("main")
+
+    # Emulate execution going to a new executor (like a thread or task).
+    provider.executor_id = 1
+    # Emulate the context being copied from the previous executor to the new
+    # one.
+    # This should be handled by contextvars (asyncio), monkeypatches (gevent)
+    # or explicit management (threads).
+    provider.activate(main_span.context)
+
+    # Close span in the wrong executor.
+    main_span.finish()
+
+    # New executor begins executing.
+    with tracer.trace("new_executor_1"):
+        pass
+
+    assert tracer.context_provider.active() is None
+
+    # provider.activate(main_span.context)
+    with tracer.trace("new_executor_2"):
+        pass
+
+    provider.executor_id = 0
+
+    with tracer.trace("main2"):
+        pass
+
+    main_span = test_spans.find_span(name="main")
+    executor_span1 = test_spans.find_span(name="new_executor_1")
+    executor_span2 = test_spans.find_span(name="new_executor_2")
+    main_span2 = test_spans.find_span(name="main2")
+
+    assert main_span.trace_id == executor_span1.trace_id
+    assert main_span.trace_id != executor_span2.trace_id
+    assert executor_span1.parent_id == main_span.span_id
+    assert executor_span2.parent_id is None
+    assert main_span2.parent_id is None
+    assert main_span2.trace_id != main_span.trace_id
+
+
+def test_default_cross_execution_parent_finish_before_child(tracer, test_spans):
+    """
+    When a trace is continued in another execution
+        When the propagated span is finished before a new span in the executor
+            The span should remain active in the executor
+    """
+
+    provider = TestContextProvider()
+    tracer.configure(context_provider=provider)
+
+    main_span = tracer.trace("main")
+
+    # Emulate execution going to a new executor (like a thread or task).
+    provider.executor_id = 1
+    # Emulate the context being copied from the previous executor to the new
+    # one.
+    # This should be handled by contextvars (asyncio), monkeypatches (gevent)
+    # or explicit management (threads).
+    provider.activate(main_span.context)
+
+    # New executor begins executing.
+    with tracer.trace("new_executor"):
+        pass
+
+    # Execution returns to main executor.
+    provider.executor_id = 0
+    main_span.finish()
+
+    main_span, executor_span = test_spans.pop()
+    assert main_span.trace_id == executor_span.trace_id
+    assert executor_span.parent_id == main_span.span_id
+
+
+def test_cross_thread_tracing(tracer, test_spans):
```

Comment:
Docstring for the test description would be great here, thanks! 😄 
i.e. 
```suggestion
def test_cross_thread_tracing(tracer, test_spans):
    ```When an active trace is continued in a new thread
         Spans created in the new thread should inherit from the previous thread.
    ```
```